lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -1263,17 +1265,42 @@ bool OverrideMatcher::checkOverride(ValueDecl *baseDecl,
     CanType parentPropertyCanTy =
       parentPropertyTy->getReducedType(
         decl->getInnermostDeclContext()->getGenericSignatureOfContext());
-    if (!propertyTy->matches(parentPropertyCanTy,
-                             TypeMatchFlags::AllowOverride)) {
-      diags.diagnose(property, diag::override_property_type_mismatch,
-                     property->getName(), propertyTy, parentPropertyTy);
-      noteFixableMismatchedTypes(decl, baseDecl);
-      diags.diagnose(baseDecl, diag::property_override_here);
-      return true;
+
+    auto options = TypeMatchFlags::AllowOverride;
+    if (!propertyTy->matches(parentPropertyCanTy, options)) {
+      // The perfect match failed
+      bool failed = true;
+      if (baseDecl->preconcurrency() &&
+          (attempt == OverrideCheckingAttempt::PerfectMatch ||
+           attempt == OverrideCheckingAttempt::MismatchedSendability)) {
```

Comment:
One major drawback of this approach is that it strips concurrency annotations from both sides which means that override with concurrency annotations would match against parent without annotation(s) in the same position which is not the case for witness matching today. For example:

```swift
protocol P {
    @preconcurrency func test(_: Any)
}

struct S : P {
    func test(_: any Sendable) {}
}
```

Produces the following error:

```
error: type 'S' does not conform to protocol 'P'
1 | protocol P {
2 |     @preconcurrency func test(_: Any)
  |                          `- note: protocol requires function 'test' with type '(Any) -> ()'; add a stub for conformance
3 | }
4 |
5 | struct S : P {
  |        `- error: type 'S' does not conform to protocol 'P'
6 |     func test(_: any Sendable) {}
  |          `- note: candidate has non-matching type '(any Sendable) -> ()'
7 | }
```

Taking a step back here, at line https://github.com/apple/swift/pull/73543/files#diff-149ea7785899898da39642522a1b7d555757f96e8a3d53f045f9ae80e04c361bR1208 the types are checked for exact match which means that it should be possible to set a new flag based on whether base is marked as preconcurrency or not before attempting `propertyTy->matches(parentPropertyCanTy, options)` and handle it inside `matches` but that would make it harder to deal with check that compares object types (that happens right under this one) and diagnose sendability mismatches as warnings, so I'd like to propose a combination:

1. We still ran `if (!propertyTy->matches(parentPropertyCanTy, options))` first;
2. If that fails we add `TypeMatchFlags::IgnoreSendability` and `TypeMatchFlags::IgnoreFunctionSendability` and re-run `matches`
3. If `propertyTy->matches(parentPropertyCanTy, options)` returns `true` we set `mismatchedOnSendability = true;` this would allow us diagnose sendability mismatch and avoid dealing with optionality mismatches diagnosed in https://github.com/apple/swift/pull/73543/files#diff-149ea7785899898da39642522a1b7d555757f96e8a3d53f045f9ae80e04c361bR1304
4. Change `matches` to handle the case where `any Sendable` is matched against `Any`  and ignore it if `IgnoreSendability` flag is set.
  - This means that `matches` has to be a bit more work by walking into tuples and collections but only if the flag is set.